doc/tsplot.1
.\" auto-generated by docbook2man-spec from docbook-utils package
.TH "TSPLOT" "1" "February 2007" "" ""
.SH NAME
tsplot \- Plot several tsung logs on the same charts, for comparison purpose.
.SH SYNOPSIS
.sp
\fBtsplot\fR [ \fB-c configuration file\fR ]  [ \fB-d images output directory\fR ]  [ \fB-v verbose\fR ]  [ \fBlegend logfile\fR ] 
.SH "DESCRIPTION"
.PP
Tsung comes with a plotting tool using
\fBgnuplot\fR, producing some graphs from the
\fItsung.log\fR file data.
\fBtsplot\fR is able to plot data from several
\fItsung.log\fR files onto the same charts serie,
for further comparison and analyze.
.SH "OPTIONS"
.PP
.TP
\fB-c\fR
.TP
\fB--config\fR
specifies the configuration file to use. Default is \fIhttp.en.plots.conf\fR\&.
.TP
\fB-d\fR
.TP
\fB--outdir\fR
directory where \fBtsplot\fR saves the images
it produces, defaults to \fI/tmp/tsung\fR\&.
.TP
\fB-v\fR
.TP
\fB--verbose\fR
makes \fBtsplot\fR very verbose about
what it does.
.SH "CONFIGURATION"
.PP
The configuration file of \fBtsplot\fR allows one to
define the plots you want to obtain, from their label to the
data they will show. The configuration file adopts a
\fI\&.ini\fR file syntax, each section defining a
chart.
.PP
\fBtsplot\fR comes with two sample configuration
files, namely \fIhttp.plots.en.conf\fR and
\fIpgsql.plots.en.conf\fR\&. They respectively define
charts to be plotted for a \fBtsung\fR HTTP test
and a \fBtsung\fR PGSQL test.
.PP
A \fBDEFAULT\fR section may be provided, any
element configured here may then be overridden into a specific
plot section.
.PP
Another configuration file is used by \fBtsplot\fR:
the \fItsung/stats.conf\fR one. It's used to
define by type the statistics to be read into
\fBtsung\fR log files, and you shouldn't need to
edit it, short of adding support for new
\fBtsung\fR statistics.
.PP
Common settings, to be found into \fBDEFAULT\fR
section or any specific chart section.
.TP
\fBencoding\fR
set here the encoding used thereafter in the file, for
labels and titles.
.TP
\fBdpi\fR
dpi setting of produced charts images
.TP
\fBtn_dpi\fR
dpi setting of produced charts thumbnail images
.TP
\fBimgtype\fR
type of chart image to produce, as in
\fIpng\fR or \fIps\fR

A complete list might be obtained on the
\fBpython-matplotlib\fR website,
http://matplotlib.sourceforge.net/
.TP
\fBxlabel\fR
default label for horizontal axe, often you want seconds
or minutes, depending on xfactor.

Please note you can also set some defaults for ylabel, but
this seems not to be a good idea in practise.
.TP
\fBxfactor\fR
tsung logs statistics in its logfile every 10 seconds. By
default, charts will not scale this and have seconds as
horizontal axis units. By setting an xfactor of 60, you
have a minute precision on horizontal axis.
.TP
\fByfactor\fR
same as xfactor, but for vertical axis.

Depending on the data you obtain with your tests, you may
want to adapt the vertical scale of your plotting. For
example, the \fBpage.mean\fR statistic is
logged in milliseconds by \fBtsung\fR\&. You
may want to display seconds if this unit better fits your
measures. Then simply set \fByfactor =
1000\fR\&.
.TP
\fBstyles\fR
set here any number of \fBmatplotlib\fR
styles you want to use, separated by spaces, as available
here:
http://matplotlib.sourceforge.net/matplotlib.pylab.html#-plot.

For exemple, set \fBstyles = b- g+ r- cx\fR
for plotting first dataset (see \fBstats\fR
below) with a blue solid line, second with green plus
symbols, third with a red line and last with cyan cross
symbols.

This could fit a \fBstats = 200.count
400.count\fR stats setting when plotting two
\fBtsung\fR logs.
.PP
You then can define any number of plot, one by section, and give
them an arbitrary name. The name must be unique, and will be
used for naming output images.
.PP
.PP
Any option available in DEFAULT section is also available in any
specific chart section, with the same meaning and effect. The
specific setting will systematically override the DEFAULT one.
.PP
.TP
\fBtitle\fR
Title of the chart, as printed into the resulting image.
.TP
\fBstats\fR
The statistics properties to use for this plotting, as
named in the \fItsung/stats\fR
configuration file. Please see this bundled file for a
list of what is available.

Tsung provide several types of statistics, as documented
here:
http://tsung.erlang-projects.org/user_manual.html#htoc53. The
two main types of statistics used are
\fBsample\fR and \fBcounter\fR\&. A
third one is \fBgauge\fR but is only use for a
single statistic (users).

sample provides count, mean, stdvar, max, min and gmean
(global mean)
properties, and counter provides only count and
totalcount. gauge provide count and max.

The stats setting can accept several
\fBstat.property\fR elements, separated by
spaces.

Examples: \fBstats = users.count\fR to plot
the number of simultaneously connected users, and
\fBstats = 200.count 400.count\fR to plot
given HTTP return codes count, both on the same chart.

Please notice \fBtsplot\fR is currently
limited to use only one horizontal and only one vertical
scales. \fBmatplotlib\fR is able to define
some more complex drawings, but \fBtsplot\fR
is not yet able to benefit from this.
.TP
\fBlegend\fR
Legend prefix, which will be followed by the legend given
on command line.

Each plot on a chart has a legend entry, you configure
here the meaning of the plot (say 'concurrent users') and
\fBtsplot\fR will add it the name of the data
serie being plotted (say 'scenario x'). You'd obtain this
legend: 'concurrent users scenario x'.
.TP
\fBylabel\fR
label for vertical axe
.SH "CONFIGURATION EXAMPLE"
.PP
Please see the given configuration examples which should be
distributed in
\fI/usr/share/doc/tsung/tsung-plotter/http.plots.en.conf\fR
and
\fI/usr/share/doc/tsung/tsung-plotter/pgsql.plots.en.conf\fR\&.
.SH "BUGS"
.PP
Please reports bugs to the mailing list
<tsung-users@process-one.net> or in the bug
tracker  <URL:https://support.process-one.net/browse/TSUN>,
see also  <URL:https://lists.process-one.net/mailman/listinfo/tsung-users> for
archives.
.SH "AUTHORS"
.PP
\fBtsplot\fR is written by Dimitri Fontaine 
<dim@tapoueh.org>\&.


CHANGES
1.4.2 -> 1.5.0 Major enhancements and bugfixes (24 May 2013)
Bugfix:
    * [TSUN-208] - in the jabber plugin, substitutions for raw request doesn't work in some cases.
    * [TSUN-209] - If tag doesn't work with Tsung 1.4.2
    * [TSUN-212] - Incorrect ERTS version being set on build.
    * [TSUN-215] - normal ack timeout shouldn't used for global ack
    * [TSUN-217] - If statement breaks on empty string
    * [TSUN-218] - Race condition in tsung-recorder
    * [TSUN-219] - Site fails to load via proxy recorder
    * [TSUN-220] - Large configuration files trigger error
    * [TSUN-229] - compatibility with erlang R15B
    * [TSUN-230] - can't connect with TLS + ejabberd
    * [TSUN-232] - Tsung for bosh protocol doesn't send a empty request to keep the user session alive.
    * [TSUN-234] - Error encoding json string with escape_uri
    * [TSUN-238] - Content-Length parsing broken
    * [TSUN-241] - Invalid link Other in the graph.html
    * [TSUN-245] - Message when dtd is not found not trivial
Improvements:
    * [TSUN-174] - add an option to set resource in XMPP
    * [TSUN-222] - Support unsubscribe operation for Jabber pubsub module
    * [TSUN-228] - allow substitutions on cookies
    * [TSUN-236] - Add probability support for servers
    * [TSUN-242] - add timestamp and request duration in dump=protocol for http
    * [TSUN-246] - http PATCH support
New Features:
    * [TSUN-214] - Ability to pass attributes for node creation for XMPP pubsub protocol.
    * [TSUN-227] - add new dynamic variable to get server hostname and port
    * [TSUN-231] - add option to use weights instead of probabilities for sessions
    * [TSUN-239] - add BOSH support
    * [TSUN-240] - add websocket support
    * [TSUN-244] - Percentile computation
    * [TSUN-248] - add AMQP support
1.4.1 -> 1.4.2 Minor enhancements and bugfixes (4 Jan 2012)
Bugfix:
   * [TSUN-199] - computation of NUsers is wrong
   * [TSUN-206] - build failure with erlang R15B
Improvements:
   * [TSUN-202] - IPv6 support
   * [TSUN-203] - snmp oids should be customizable in the config file
   * [TSUN-205] - handle dyn_variables as array in test conditions (if/until/while)
New Features:
   * [TSUN-191] - allow outputting log to stdout
   * [TSUN-192] - structured log output (JSON)
   * [TSUN-193] - accept configuration from stdin
   * [TSUN-197] - Have bug and error message on stderr and not stdout.
1.4.0 -> 1.4.1 Minor bugfixes (13 Sep 2011)
Bugfix:
   * [TSUN-188] - munin plugin is not working in 1.4.0
   * [TSUN-189] - the controller VM is not used in some case
   * [TSUN-190] - pgsql recorder can record a connect request in an already connected session
1.3.3 -> 1.4.0 Major enhancements and bugfixes (5 Sep 2011)
Bugfix:
   * [TSUN-129] - regexp (defined in match or dynvars) can fail when chunk encoding is used.
   * [TSUN-150] - Munin monitoring broken by cpu stats config request
   * [TSUN-163] - Tsung doesn't detect subdomains.
   * [TSUN-166] - snmp monitoring does not work with erlang R14A
   * [TSUN-171] - maxnumber set in a phase is not always enforced
   * [TSUN-172] - auth sasl can't authenticate against ejabberd
   * [TSUN-178] - some characters can make url and headers rewriting fail in the recorder
   * [TSUN-179] - tsung generated message stanzas are not XMPP compliant
   * [TSUN-180] - file server crash if a dynamic substitution use it
   * [TSUN-182] - When many clients are configured with few static users, none of them are launched.
   * [TSUN-183] - tsung can stop too soon in some cases
   * [TSUN-184] - 'random_number' with start and end actually returns a number from start+1 to end
   * [TSUN-187] - Client IP scan is very slow on Linux; also uses obsolete "ifconfig"
Improvements:
   * [TSUN-54] - tsung is very slow when a lot of dynamic variables are set
   * [TSUN-96] - generating more than 64k connections from a single machine
   * [TSUN-106] - Add content-encoding support for dynvar extraction
   * [TSUN-123] - add option to read usernames from an external file for jabber
   * [TSUN-125] - use the new re module everywhere instead of regexp/gregexp
   * [TSUN-152] - Add "state_rcv" record as parameter to "get_message" function.
   * [TSUN-153] - dynvar used in match may contain regexp special characters
   * [TSUN-185] - handle postgresql extended protocol
New Features:
   * [TSUN-162] - add foreach tag (loop when a dyn_variable is a list)
   * [TSUN-164] - add a switch to allow light queries/replies logging
   * [TSUN-165] - add a way to synchronize users for all plugins.
   * [TSUN-167] - add do=dump option to matching
   * [TSUN-168] - thinktimes value could be dynamically generated with dyn_variable
   * [TSUN-181] - add option to simulate slow connections
1.3.2 -> 1.3.3 Minor bugfixes (17 Aug 2010)
Bugfix:
    * [TSUN-154] - parent proxy doesn't work anymore in 1.3.x (tested with 1.3.2 and 1.3.0).
    * [TSUN-155] - url substitution is broken in some cases
    * [TSUN-156] - Tsung not using sessions with low probabilities
    * [TSUN-157] - ssl doesn't work with erlang R14A
    * [TSUN-158] - failure when a proxy is used and an URL substitution is set
    * [TSUN-159] - HTTP cookies support is broken when a proxy is used
    * [TSUN-160] - tsung can sometimes hang at the beginning using distributed setup
    * [TSUN-161] - if statement not allowed in a transaction
1.3.1 -> 1.3.2 Major bugfixes and enhancements (14 Jun 2010)
Bugfix:
    * [TSUN-128] - Apostrophes cause string to convert to deep list in setdynvars with Erlang function.
    * [TSUN-130] - static users starting time is wrong
    * [TSUN-131] - tsung can stop too early when static users are used
    * [TSUN-132] - http cookies: accept domains equals to hostname with a leading "."
    * [TSUN-133] - proxy-recorder with SSL fails on large client packets (multiple TCP packets)
    * [TSUN-138] - when an error occured( for ex a timeout during a request) and a client exits, started transactions are not updated

src/tsung-plotter/tsplot.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

#
#  Created: 2006 by Dimitri Fontaine <dim@tapoueh.org>
#
#  Modified by Nicolas Niclausse
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
#
#  In addition, as a special exception, you have the permission to
#  link the code of this program with any library released under
#  the EPL license and distribute linked combinations including
#  the two.

# A plotter for tsung text generated data

from tsung.tsung import TsungLog
import os, sys
from ConfigParser import ConfigParser
from pylab import *

SYSDIR         = "./"
USERDIR        = ".tsung"
CONF           = "plots.conf"
SYS_STATS_CONF = os.path.join(SYSDIR, 'tsung', 'stats.conf')

class Plot:
    """ The ploting class, using matplotlib """

    def __init__(self,
                 name    = "",
                 xlabel  = "",
                 ylabel  = "",
                 title   = "",
                 legends = None,
                 styles  = ['b-', 'r-', 'g.', 'c,', 'y^', 'kv'],
                 dpi     = 150,
                 tn_dpi  = 50,
                 xfactor = 1,
                 yfactor = 1,
                 outdir  = '/tmp/tsung-plotter',
                 imgtype = 'png'):

        # Only used if no plot type given
        self.name      = name
        self.xlabel    = xlabel
        self.ylabel    = ylabel
        self.title     = title
        self.legends   = legends
        self.dpi       = dpi
        self.thumb_dpi = tn_dpi
        self.xfactor   = xfactor
        self.yfactor   = yfactor
        self.styles    = styles
        self.outdir    = outdir
        self.imgtype   = imgtype

    def plot(self, stats, dataset):
        """ draw a simple timeline plots from two dataset """

        # prepare matplotlib graph
        clf()
        ax = subplot(111)

        count = 0
        for data in dataset:
            # get give type data for plotting
            for (name, stat) in stats:
                pdata = data.stat(name, stat)

                # we want timestamp sorted data to plot
                ts = pdata.keys()
                ts.sort()
                values = [pdata[t] / self.yfactor for t in ts]

                if self.xfactor not in [1, "1"]:
                    ts = [x / self.xfactor for x in ts]

                # Now use matplotlib to do the plotting
                plot(ts, values, self.styles[count%len(self.styles)])
                count += 1

        # Now setup the legend
        title(self.title)
        xlabel(self.xlabel)
        ylabel(self.ylabel)
        legend(self.legends)

        # we want to draw a grid
        grid(True)

        filename = os.path.join(self.outdir,
                                "%s.%s" % (self.name, self.imgtype))
        thumbnail = os.path.join(self.outdir,
                                 "%s_tn.%s" % (self.name, self.imgtype))

        savefig(filename, dpi=self.dpi, orientation='portrait')
        savefig(thumbnail, dpi=self.thumb_dpi, orientation='portrait')

        return filename


def main(conffile, logs, legends, outdir, verbose):
    """ Produce plots from given """

    dataset = [d for f, d in logs]

    config = ConfigParser()
    config.read(conffile)

    for s in config.sections():
        p = Plot(name = s, outdir = outdir)

        # defaults
        for d, v in config.defaults().items():
            if d != 'encoding':
                p.__dict__[d] = v
            else:
                encoding = v

        # stats
        # conf:   200.count, 400.count
        # result: [["200", "count"], ["400", "count"]]
        try:
            stats = [x.strip().split('.')
                     for x in config.get(s, 'stats').split(' ')]
        except:
            print 'error: unable to read plot "%s" stats' % s
            continue

        if config.has_option(s, 'styles'):
            p.styles = [x.strip() for x in config.get(s, 'styles').split(' ')]

        if config.has_option(s, 'legend'):
            # this is the legend prefix$
            l = []
            count    = 0
            clegends = config.get(s, 'legend').decode(encoding).split(',')

            for f in logs:
                l     += ['%s %s' % (x.strip(), legends[count]) \
                          for x in clegends]
                count += 1

        p.legends = l

        # Text parameters - to decode into specified encoding
        for attr in ['title', 'xlabel', 'ylabel']:
            if config.has_option(s, attr):
                cstring = config.get(s, attr).decode(encoding)
                p.__dict__[attr] = cstring

        # Numerical parameters
        for attr in ['xfactor', 'yfactor', 'dpi', 'tn_dpi']:
            if config.has_option(s, attr):
                try:
                    p.__dict__[attr] = config.getfloat(s, attr)
                except ValueError:
                    print 'warning: %s %s not a number: %s' \
                          % (p.name, attr, config.get(s, attr))

        outfile = p.plot(stats, dataset)

        if verbose:
            print 'Generated plot %s' % outfile

if __name__ == "__main__":
    from optparse import OptionParser

    parser = OptionParser()
    parser.add_option("-c", "--config", dest="config", default=None,
                      help="configuration file")
    parser.add_option("-d", "--outdir", dest="outdir", default="/tmp/tsung",
                      help="output dir where to save plots (/tmp/tsung)")
    parser.add_option("-v", action="store_true", dest="verbose", default=False,
                      help="be verbose")

    (options, args) = parser.parse_args()

    if options.config is None:
        userconf = os.path.join(os.environ['HOME'], USERDIR, CONF)
        if os.access(userconf, os.R_OK):
            config = userconf
        else:
            config = os.path.join(SYSDIR, CONF)
    else:
        config = options.config

    if options.verbose:
        print 'Using %s configuration file' % config

    if not os.access(config, os.R_OK):
        print "can't read configuration file: %s" % config
        sys.exit(1)

    # FIXME: error control
    # OSError: [Errno 17] Le fichier existe.: '/tmp/tsung'
    try:
        os.makedirs(options.outdir)
    except:
        pass

    # args are legend then file, any times wanted by user
    if len(args) % 2 != 0:
        print "error: please provide legend and tsung log filename"
        sys.exit(3)

    count   = 0
    legends = []
    files   = []

    for a in args:
        if count % 2 == 0:
            legends.append(a)
        else:
            files.append(a)

        count += 1

    if options.verbose:
        print 'Using %s stats configuration file' % SYS_STATS_CONF

    logs = []
    for logfile in files:
        if not os.access(logfile, os.R_OK):
            print "error: unable to read file %s" % logfile

        else:
            if options.verbose:
                print 'Parsing Tsung log file', logfile
            logs.append((logfile, TsungLog(SYS_STATS_CONF, logfile)))

    if len(logs) != len(args) / 2:
        print 'error while parsing files (%d != %d)' % (len(logs),
                                                        len(args)/2)
        sys.exit(2)

    main(config, logs, legends, options.outdir, options.verbose)
